### PR TITLE
Tandem Auxiliary Heads: stagnation/suction peak location prediction

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Add **two lightweight auxiliary prediction heads** that explicitly predict key aerodynamic landmarks on the aft-foil surface: (1) the stagnation point location and (2) the suction peak location/magnitude. These auxiliary tasks force the backbone to develop representations that encode the critical physical features driving tandem interference — the stagnation point shift (caused by downwash from the fore-foil) and the suction peak (which determines the adverse pressure gradient and hence separation).

**Why this helps p_tan:** The dominant error in tandem pressure prediction comes from two sources: (a) wrong stagnation point location on the aft-foil (shifted by inter-foil downwash), and (b) wrong suction peak magnitude/location (affected by gap/stagger). By adding auxiliary heads that EXPLICITLY predict these landmarks, the backbone is forced to learn representations that encode the tandem interference signal. This is an auxiliary LOSS approach (proven successful: Edward's Kutta constraint #2374 shows p_oodc beating baseline), not a conditioning approach (which failed in your Cl/Cd experiment).

**Key difference from Cl/Cd conditioning (#2363):** This approach uses auxiliary PREDICTION heads (gradient flows INTO the backbone, improving representations) rather than CONDITIONING (which feeds noisy scalars into the decoder). No information bottleneck — the backbone learns rich representations rather than consuming lossy integrals.

### Architecture

Add two small auxiliary MLP heads that predict per-foil aerodynamic landmarks:

**1. Stagnation Point Head:**
- Input: pooled backbone hidden states for each foil (mean of surface node hidden states)
- Architecture: `Linear(192, 64) → GELU → Linear(64, 1)` — predicts arc-length position of stagnation point (Cp_max location)
- Target: arc-length coordinate of the node with maximum Cp in the ground truth surface pressure
- Loss: L1 loss, weight 0.05
- Applied to BOTH foils in tandem and to single-foil samples

**2. Suction Peak Head:**
- Input: same pooled backbone hidden states per foil
- Architecture: `Linear(192, 64) → GELU → Linear(64, 2)` — predicts (arc_length_position, magnitude) of the suction peak (Cp_min location)
- Target: (arc-length coordinate, Cp value) of the node with minimum Cp in the ground truth
- Loss: L1 loss on both outputs, weight 0.05
- Applied to BOTH foils

### Implementation Details

Add flag: `--tandem_aux_heads`

When active:
1. During data loading, compute targets for each sample/foil:
   - `stag_pos` = arc-length of node with max(Cp) on each foil
   - `suction_pos` = arc-length of node with min(Cp) on each foil
   - `suction_mag` = min(Cp) value on each foil
   - Normalize arc-length to [0, 1] range per foil
2. After the backbone forward pass, pool surface node hidden states per foil: `h_foil = mean(h[surface_nodes_foil_k])`
3. Run both heads on `h_foil` for each foil
4. Compute auxiliary losses and add to total loss:
   ```python
   L_stag = 0.05 * L1(stag_pred, stag_target)
   L_suction = 0.05 * L1(suction_pred, suction_target)
   L_total = L_main + L_stag + L_suction
   ```
5. These auxiliary losses should be added to the PRESSURE task head in PCGrad (they share the same physical objective)

### Key points:
- The auxiliary heads are TINY (~13K params total) — negligible compute/memory overhead
- The gradient from these heads flows back through the backbone's surface node representations, enriching them
- For tandem: the aft-foil stagnation point shifts significantly with gap/stagger. Predicting this correctly forces the backbone to encode the tandem interference explicitly
- These targets (max/min Cp locations) are deterministic from the ground truth — no noise, no approximation
- At inference time, the auxiliary heads are unused (but can provide useful diagnostics)

### Training Setup

```bash
cd cfd_tandemfoil && python train.py \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only \
  --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity \
  --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --tandem_aux_heads \
  --wandb_group tandem-aux-heads \
  --seed 42
```

Run 2 seeds (42, 73) as standard.

### What to Watch

- **p_tan** is the primary target — the auxiliary heads specifically encode tandem interference landmarks
- **p_in** should NOT regress — the heads also operate on single-foil (stagnation/suction points exist for all airfoils)
- Log auxiliary head losses separately: `stag_loss`, `suction_loss` — they should decrease during training
- Check that the stagnation point prediction converges (L1 < 0.05 in normalized arc-length) — if it doesn't, the backbone isn't learning the interference signal
- Epoch time overhead should be < 2% (tiny auxiliary heads)

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.872** | < 11.872 |
| p_oodc | 7.459 | < 7.459 |
| **p_tan** | **26.319** | < 26.319 |
| **p_re** | **6.229** | < 6.229 |

Baseline PR: #2357 (Vortex-Panel Induced Velocity)

Reproduce baseline:
```
cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2383 because ADVISOR: Closing — aux heads didn't beat baseline overall. EMA 2-seed avg: p_in=11.71 (-1.4% beats!), but p_oodc=7.79 (+4.5%), p_tan=27.53 (+4.6%), p_re=6.59 (+5.8%). W&B: t0ewyf3m (s73), h93n8vc9 (s42).

The stagnation/suction peak location prediction heads added some useful gradient signal for p_in but created interference for OOD metrics. The auxiliary losses competing with the main surface MAE loss likely destabilized the PCGrad balancing.

Moving you to a new experiment — Bernoulli velocity-pressure consistency constraint..

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Tandem Auxiliary Heads — Stagnation/Suction Peak Location Prediction

| Metric | Baseline (2-seed avg) | Ours (2-seed avg) | Delta | Seed 42 | Seed 73 |
|--------|----------------------|-------------------|-------|---------|---------|
| **p_in** | **11.872** | **11.700** | **-0.172** | 11.9 | 11.5 |
| p_oodc | 7.459 | 7.800 | +0.341 | 7.9 | 7.7 |
| **p_tan** | **26.319** | **27.550** | **+1.231** | 27.6 | 27.5 |
| p_re | 6.229 | 6.600 | +0.371 | 6.6 | 6.6 |

### Training Details
- **Command**: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 --tandem_aux_heads --wandb_group tandem-aux-heads --seed {42,73}`
- **W&B runs**: `h93n8vc9` (seed 42), `t0ewyf3m` (seed 73)
- **Peak VRAM**: ~40.6 GB (negligible overhead from aux heads)
- **Epoch time**: ~76s (< 2% overhead vs baseline)
- **Best epoch**: 142 (s42), 143 (s73)
- **Auxiliary head params**: 24,899 (~0.002% of total)

### What Happened

**The hypothesis did not beat the baseline.** While p_in improved slightly (-0.17), all other metrics regressed — most notably p_tan (+1.23), which was the primary target.

**Analysis:**

1. **p_tan regressed (+1.23)**: This is the opposite of what was expected. The auxiliary heads were specifically designed to improve tandem pressure prediction by forcing the backbone to encode stagnation point shifts and suction peak variations. Instead, the additional loss terms may have interfered with the delicate gradient balance maintained by PCGrad 3-way.

2. **p_oodc and p_re also regressed**: The auxiliary losses are shared across all PCGrad groups (added to each group's loss). This effectively increases the weight of the shared loss terms relative to the group-specific surface/volume losses, potentially diluting the PCGrad gradient surgery that helps OOD generalization.

3. **Gradient interference with PCGrad**: The auxiliary heads share the same backbone hidden states as the main task. In the PCGrad path, the auxiliary losses are added to every group, meaning their gradients appear (and get projected) in every group. This changes the effective gradient direction compared to the baseline, even though the auxiliary objectives are aligned with pressure prediction.

4. **Mean-pooled representations may be too lossy**: The auxiliary heads receive mean-pooled hidden states across all surface nodes of a foil. This throws away the spatial structure that distinguishes stagnation points from suction peaks. A more effective approach might use attention-weighted pooling or direct per-node predictions.

5. **The suction peak Cp magnitude target has high variance**: Cp_min ranges from -0.5 to -10+ across samples, causing the suction loss magnitude component to dominate and potentially inject noisy gradients into the backbone.

### Suggested Follow-ups

1. **Try without PCGrad integration**: Add aux losses only in the non-PCGrad fallback path (when all samples are same type). This preserves the PCGrad gradient surgery while still training the auxiliary heads.
2. **Reduce aux weight**: Try 0.01 instead of 0.05 to minimize interference with the main loss landscape.
3. **Normalize suction magnitude**: Scale the Cp target to [0, 1] to prevent magnitude domination.
4. **Attention-weighted pooling**: Instead of mean-pooling, use a small attention layer to weight surface nodes, giving more signal to the stagnation/suction regions.
5. **Auxiliary heads on aft-foil only**: Since p_tan is the target, restrict aux heads to aft-foil surface nodes only.
